src/js/sonifier.js

```diff
@@ -0,0 +1,402 @@
+                gradeNames: ["floe.chartAuthoring.electricPianoBand"],
+
+                components: {
+                    scheduler: {
+                        type: "flock.scheduler.async",
+                        options: {
+                            synthContext: "floe.chartAuthoring.electricPianoBand"
+                        }
+                    }
+                }
+            });
+
+            var dataPianoBand = floe.chartAuthoring.dataPianoBand();
+
+            // Add the synth to the model
+            that.applier.change("synth", dataPianoBand);
+
+            var sonifiedData = that.model.sonifiedData;
+            // Copy the sonification definition into the queue
+            that.applier.change("sonificationQueue",sonifiedData);
+        }
+    };
+
+    // Passed a sonified dataset, this function + playDataAndQueueNext acts recursively
+    // to loop through the dataset, play a voice label + sonified data, then
+    // schedule the next play event based on the timing
+
+    // The "delay" variable manages the time to elapse (in seconds) before
+    // beginning the voice label + sonification of the particular datapoint.
+    // This is required for scheduling with the Flocking-based synth generator,
+    // which plays asynchronously without callbacks, while voice events have
+    // callbacks to indicate their speech has completed
+    //
+    // The basic challenge of scheduling that this approach handles is:
+    // - we can calculate in advance how long the sonification of a data point
+    // takes by summing note duration, but we can't receive event notification
+    // when a sonification completes
+    // - we can fire an event when a voice label read completes, but can't know
+    // in advance how long it will take to read the label
+    floe.chartAuthoring.sonifier.processSonificationQueue = function(delay, noGap, that) {
+        that.applier.change("isPlaying", true);
+        var gapDuration = that.options.playbackOptions.gapDuration;
+
+        var sonificationQueue = that.model.sonificationQueue;
+
+        var synth = that.model.synth;
+
+        var currentData = sonificationQueue[0];
+
+        var textToSpeech = that.textToSpeech;
+
+        // Schedule the next voice label, accounting for both the variable-length
+        // delay (the time to play the preceding sonification) and the fixed-length
+        // gap
+
+        // We shouldn't use the gap if this is the first call for a dataset
+        var pause = noGap? delay : delay+gapDuration;
+        if(fluid.textToSpeech.isSupported()) {
+            synth.scheduler.once(pause, function() {
+                that.applier.change("currentlyPlayingData", currentData);
+                textToSpeech.queueSpeech(currentData.label);
+            });
+        }
+    };
+
+    // Recursion function called from floe.chartAuthoring.sonifier.processSonificationQueue
+
+    floe.chartAuthoring.sonifier.playDataAndQueueNext = function(that) {
```

Comment:
@amb26 just speculative, but I imagine a fully declarative scheduler might look similar to `fluid.test.testCaseHolder` - they're both handling some similar challenges of ensuring a series of mixed synchronous and asynchronous events happen in a certain sequence. That might be wildly off base, but it's a thought that occurs to me.

I'd characterize what I've done here as a first pass at a well-behaved scheduler - still too imperative, but hopefully going in the right direction. I'd eventually (maybe) see something like a scheduler grade that works in a similar manner to the queue processing approach I've used here, where a collection of objects with attached behaviours is fed one by one into a processing queue through a combination of events and listeners - hence why I say this might look similar to `fluid.test.testCaseHolder`.